saleor/payment/tests/test_payment.py

```diff
@@ -550,3 +562,237 @@ def test_update_payment(gateway_response, payment_txn_captured):
+        (
+            ["2021-05-01 12:00:00.0+00:00"],
+            False,
+            True,
+            True,
+            True,
+            False,
+        ),  # not release, not valid kind
+        (
+            ["2021-05-01 12:00:00.0+00:00"],
+            False,
+            True,
+            False,
+            False,
+            False,
+        ),  # not release, not success
+        (
+            ["2021-05-01 12:00:00.0+00:00"],
+            False,
+            True,
+            False,
+            True,
+            True,
+        ),  # not release, action required
+    ],
+)
+def test_payment_not_needed_to_release(
+    payment_dummy,
+    dates,
+    order_exist,
+    is_active,
+    not_valid_kind,
+    success,
+    action_required,
+):
+    payment = payment_dummy
+    if not order_exist:
+        payment.order = None
+        payment.save()
+
+    if not is_active:
+        payment.is_active = False
+        payment.save()
+
+    for date in dates:
+        with freeze_time(date):
+            kind = (
+                TransactionKind.CAPTURE
+                if not not_valid_kind
+                else TransactionKind.CANCEL
+            )
+            Transaction.objects.create(
+                payment=payment,
+                amount=payment.total,
+                kind=kind,
+                gateway_response={},
+                is_success=success,
+                action_required=action_required,
+            )
+
+    payments = get_unfinished_payments()
+    assert payments.count() == 0
+    assert not is_payment_unfinished_and_ready_to_release(payment)
+
+
+@mock.patch(
+    "saleor.payment.gateway.payment_refund_or_void",
+)
+@mock.patch(
+    "saleor.payment.utils.is_payment_unfinished_and_ready_to_release",
+)
+def test_payment_release(
+    is_payment_unfinished_and_ready_to_release,
+    payment_refund_or_void,
+    payment_dummy,
+    checkout_with_item,
+):
+    payment = payment_dummy
+    payment.checkout = checkout_with_item
+    payment.save()
+
+    is_payment_unfinished_and_ready_to_release.return_value = True
+
+    manager = get_plugins_manager()
+    release_checkout_payment(payment, manager)
+    payment_refund_or_void.assert_called_once()
+
+
+@mock.patch(
+    "saleor.payment.gateway.payment_refund_or_void",
+)
+@mock.patch(
+    "saleor.payment.utils.is_payment_unfinished_and_ready_to_release",
+)
+@pytest.mark.parametrize(
+    "should_be_released, has_checkout",
+    [(False, True), (False, False)],
+)
+def test_payment_not_release(
+    is_payment_unfinished_and_ready_to_release,
+    payment_refund_or_void,
+    payment_dummy,
+    checkout_with_item,
+    should_be_released,
+    has_checkout,
+):
+    payment = payment_dummy
+    if has_checkout:
+        payment.checkout = checkout_with_item
+        payment.save()
+
+    is_payment_unfinished_and_ready_to_release.return_value = should_be_released
+
+    with pytest.raises(ReleasePaymentException):
+        manager = get_plugins_manager()
```

Comment:
IMHO, for better readability, you should place in `with pytest.rais...` only a part of code which can raise the error. 
Manager and mock should be excluded from it.

Reply:
OK